### PR TITLE
chore: versioneer-compatible versions

### PR DESCRIPTION
**Pull request by aarnphm**
Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

Maybe i spent way too much time writing stubs :)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -112,7 +112,7 @@ skip_glob = ["typings"]
 
 [tool.pyright]
 include = ['bentoml/**/*']
-exclude = ['bentoml/_version.py','bentoml/__main__.py','bentoml/testing/**/*','tests/**/*','typings/**/*','docker/**/*']
+exclude = ['bentoml/_version.py','versioneer.py','bentoml/__main__.py','bentoml/testing/**/*','tests/**/*','typings/**/*','docker/**/*']
```
Comment (parano) [inline pyproject.toml:115]: revert this change?

File: bentoml/_internal/configuration/__init__.py [hunk 2]
```diff
@@ -42,7 +42,7 @@ def expand_env_var(env_var):
 
 
 def is_pip_installed_bentoml():
-    is_installed_package = hasattr(version_mod, "version_json")
+    is_installed_package = len(version_mod.version_tuple) == 3
```
Comment (parano) [inline bentoml/_internal/configuration/__init__.py:45]: we will start with pre-release on pypi which will break this, pre-release looks like `1.0.0.a2`

Comment (aarnphm) [reply bentoml/_internal/configuration/__init__.py:45]: :) it shouldn't

Comment (parano) [reply bentoml/_internal/configuration/__init__.py:45]: tried building a pre-release with `1.0.0.a2` and `bentoml._version.version_tuple == (1,0,0)` after `pip install`, so it's working as expected.

I do run into another case, where after I tag the branch with any string in semver format, it will think it's a pypi released version, even if it's a editable install

File: bentoml/_internal/configuration/__init__.py [hunk 3]
```diff
@@ -42,7 +42,7 @@ def expand_env_var(env_var):
...
     is_tagged = not BENTOML_VERSION.startswith("0+untagged")
```
Comment (parano) [inline bentoml/_internal/configuration/__init__.py:46]: we will also need to change this to `not BENTOML_VERSION.startswith("0.1.")` - looks like setuptools_scm will generate versions like `0.1.dev1392+gddbebc4` when there is no tag

Comment (aarnphm) [reply bentoml/_internal/configuration/__init__.py]: done

File: MANIFEST.in [hunk 4]
```diff
@@ -2,6 +2,10 @@ include README.md
 include LICENSE
 
 # Included in PyPI distribution
+include bentoml/*.py
```
Comment (parano) [inline MANIFEST.in:5]: @aarnphm these lines here are not needed

Comment (parano) [reply MANIFEST.in:5]: all files tracked by git is already included by default,  now with setuptools scm

Comment (aarnphm) [reply MANIFEST.in:5]: noted

File: bentoml/_internal/configuration/__init__.py [hunk 5]
```diff
@@ -2,14 +2,23 @@
...
 except ModuleNotFoundError:
     import importlib_metadata
 
 import yaml
 
-import bentoml._version as version_mod
+try:
+    import bentoml._version as version_mod
+except ImportError:
+    # Since we aren't VCS _version.py, which is generated by setuptools_scm
```
Comment (parano) [inline bentoml/_internal/configuration/__init__.py:17]: when will this happen? 

I suppose only when a developer does `pip install -e .` and then removed the `_version.py` file by accident?

Comment (aarnphm) [reply bentoml/_internal/configuration/__init__.py:17]: yes. Also mainly for type checking where pyright cannot find `bentoml._version`.
There is also a current caveats with types CI atm

File: bentoml/_internal/configuration/__init__.py [hunk 6]
```diff
@@ -41,22 +52,23 @@ def expand_env_var(env_var):
...
     is_modified = BENTOML_VERSION != BENTOML_VERSION.split("+")[0]
     return is_installed_package and is_tagged and is_clean and is_modified
 
 
-def get_bentoml_config_file_from_env():
+def get_bentoml_config_file_from_env() -> t.Optional[str]:
     if CONFIG_ENV_VAR in os.environ:
         # User local config file for customizing bentoml
-        return expand_env_var(os.environ.get(CONFIG_ENV_VAR))
+        return expand_env_var(os.environ[CONFIG_ENV_VAR])
```
Comment (parano) [inline bentoml/_internal/configuration/__init__.py:67]: this will raise a keyerror if the env var is not set

File: bentoml/_internal/configuration/containers.py [hunk 7]
```diff
@@ -62,17 +68,15 @@
...
             ),
             "console": {"enabled": bool},
             "file": {"enabled": bool, "directory": Or(str, None)},
             "advanced": {"enabled": bool, "config": Or(dict, None)},
         },
         "tracing": {
-            "type": Or(
-                And(str, Use(str.lower), lambda s: s in ("zipkin", "jaeger")), None
-            ),
+            "type": Or(And(str, Use(str.lower), _check_tracing_type), None),
```
Comment (parano) [inline bentoml/_internal/configuration/containers.py:79]: this case is better left it is, easier to read in context

Comment (aarnphm) [reply bentoml/_internal/configuration/containers.py:79]: This mainly to compensate pyright :(

File: bentoml/_internal/configuration/containers.py [hunk 8]
```diff
@@ -219,14 +223,17 @@ def tracer(
 
             return NoopTracer()
 
-    logging_file_directory = providers.Factory(
-        lambda default, customized: customized or default,
+    logging_file_directory = t.cast(
```
Comment (parano) [inline bentoml/_internal/configuration/containers.py:226]: @bojiang could you review this part?

File: scripts/ci/style/pylint_check.sh [hunk 9]
```diff
@@ -4,17 +4,27 @@ name=$(basename "$0")
...
-pylint --rcfile="./pyproject.toml" --disable=E0401,F0010 tests docker | tee /tmp/"$name"_tests_docker
+if [[ -n "$GITHUB_BASE_REF" ]]; then
+  echo "Running pylint on changed files..."
+  git fetch origin "$GITHUB_BASE_REF"
+  if ! (git diff --name-only --diff-filter=d "origin/$GITHUB_BASE_REF" -z -- '*.py' | xargs -0 --no-run-if-empty pylint --rcfile="$GIT_ROOT/pyproject.toml" --exit-zero); then
+    FAIL "pylint failed."
+    exit 1
+  fi
+else
+  echo "Running pylint for the whole library..."
```
Comment (parano) [inline scripts/ci/style/pylint_check.sh:19]: also I think pylint is failing on main branch now,  for the same reason pyright was failing before

Comment (aarnphm) [reply scripts/ci/style/pylint_check.sh:19]: We can just disable pylint for running on the main branch then

File: bentoml/_internal/configuration/containers.py (hunk 8 above)
Comment (bojiang) [reply bentoml/_internal/configuration/containers.py:226]: what?

Comment (bojiang) [reply bentoml/_internal/configuration/containers.py:226]: lol it should not be like this

File: bentoml/_internal/configuration/containers.py (hunk 7 above)
Comment (parano) [reply bentoml/_internal/configuration/containers.py:79]: can't believe there's no easy way to type a lambda function inline :(

File: bentoml/_internal/configuration/containers.py (hunk 8 above)
Comment (aarnphm) [reply bentoml/_internal/configuration/containers.py:226]: simple_di can't be inferred by pyright

File: bentoml/_internal/configuration/__init__.py (hunk 6 above)
Comment (parano) [reply bentoml/_internal/configuration/__init__.py:67]: could you also make this change in the follow-up pr?
